### PR TITLE
Extend type support on JWT claims

### LINKED ISSUE #439: Support Map as an additional claim on JWTClaim payload
Currently `JwtClaims` only support String objects when passing a map of additional claims. This limitation prevent applications that need more sophisticated JWT payloads to use the library.

According to the [RFC 7519](https://tools.ietf.org/html/rfc7519) a claim value can be any valid JSON value including numbers, booleans, dates, list and other objects.

An example of this feature being implemented can be found on this [commit](https://github.com/auth0/java-jwt/commit/343c3a2454f889c399ceed755bb5f6d33f2e2496?diff=unified) for the library `java-jwt`, by auth0.

#### Code snippet

This is how usage would look like:

```java
Map<String, String> value = ImmutableMap.of("foo", "bar");

Map<String, String> stringClaim = ImmutableMap.of("fooStr", "barStr");
Map<String, Map<String, ?>> mapClaim = 
    Collections.<String, Map<String, ?>>singletonMap("zaz", value);

JwtClaims claims = JwtClaims.newBuilder()
        .setAdditionalClaims(stringClaim)
        .setAdditionalComplexClaims(mapClaim) // [NEW]
        .build();
```

1. Note that the introduction of another setter instead of changing `setAdditionalClaims` signature is to keep backwards compatibility. The newly introduced `setAdditionalComplexClaims` could support either a Map or a wild card (`?`) type, with validations happening during the `build()` call. I am open to both approaches and any recommendation would be welcome.

1. Also not that because of the above a change on `JwtCredentials` is also required replacing the following code when generating the JWT payload:

```java
// used to be  payload.putAll(jwtClaims.getAdditionalClaims());
payload.putAll(jwtClaims.getAllAdditionalClaims());
```
1. `getAllAdditionalClaims` would be a map combining both `getAdditionalClaims` and `getAdditionalComplexClaims`. This would allow current users who set their claims through the original method to keep their code functional.

#### External references such as API reference guides used

- [RFC 7519](https://tools.ietf.org/html/rfc7519)
- [java-jwt library implementation](https://github.com/auth0/java-jwt/commit/343c3a2454f889c399ceed755bb5f6d33f2e2496?diff=unified) 

#### Any additional information below

Should this issue be accepted I'm willing to move forward with the implementation suggested above and create a Pull Request contributing to the library. Please also advice if there's any recommendation for the item (2) on the list above.

Thanks!

### PR DESCRIPTION
**Pull request by rafaelmenta**
Additional JWT claims now support more types other than String

Fixes #439

### CODE REVIEW

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java [hunk 1]
```diff
@@ -111,14 +118,111 @@ public boolean isComplete() {
...
+        }
+      }
+
+      return true;
+    }
+
+    /**
+     * Validates if a collection is a valid JSON value. Empty collections are considered valid.
+     *
+     * @param collection collection to be evaluated.
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:202]: nit: no period since not a complete sentence

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java [hunk 2]
```diff
@@ -111,14 +118,111 @@ public boolean isComplete() {
...
+        return validateKeys((Map) object) && validateCollection(((Map) object).values());
+      }
+
+      return isSupportedValue(object);
+    }
+
+    /**
+     * Validates the keys on a given map. Keys must be Strings.
+     *
+     * @param map map to be evaluated.
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:187]: nit: no period since not a complete sentence

File: oauth2_http/java/com/google/auth/oauth2/JwtCredentials.java [hunk 3]
```diff
@@ -34,6 +34,7 @@
 import com.google.api.client.json.webtoken.JsonWebSignature;
 import com.google.api.client.json.webtoken.JsonWebToken;
 import com.google.api.client.util.Clock;
+import com.google.api.client.util.GenericData;
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/JwtCredentials.java:37]: why is this here if there are no other changes in this class?

Comment (rafaelmenta) [reply oauth2_http/java/com/google/auth/oauth2/JwtCredentials.java:37]: Leftover from some experimentation I was doing before, removed.

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java (hunk 1 above)
Comment (rafaelmenta) [reply oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:202]: Fixed, thanks

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java (hunk 2 above)
Comment (rafaelmenta) [reply oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:187]: Fixed, thanks

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java [hunk 4]
```diff
@@ -66,12 +71,13 @@
   /**
    * Returns additional claims for this object. The returned map is not guaranteed to be mutable.
    *
-   * @return additional claims
+   * @return additional string claims
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:74]: I'm confused here. The type no longer says it's a string but the doc comment says it is?

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java [hunk 5]
```diff
@@ -111,14 +118,111 @@ public boolean isComplete() {
...
+      Preconditions.checkState(validateClaims(claims.getAdditionalClaims()), ERROR_MESSAGE);
+      return claims;
+    }
+
+    /**
+     * Validate if the objects on a Map are valid for a JWT claim.
+     *
+     * @param claims Map of claim objects to be validated
+     */
+    private static boolean validateClaims(Map<String, ?> claims) {
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:150]: This is interesting code. I wonder if there's a library for this somewhere?

File: oauth2_http/javatests/com/google/auth/oauth2/JwtClaimsTest.java [hunk 6]
```diff
@@ -31,11 +31,15 @@
 
 package com.google.auth.oauth2;
 
+import static com.google.common.collect.ImmutableMap.builder;
 import static org.junit.Assert.*;
 
-import java.util.Collections;
-import java.util.Map;
+import java.util.*;
```
Comment (elharo) [inline oauth2_http/javatests/com/google/auth/oauth2/JwtClaimsTest.java:37]: no wildcard imports per google style

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java (hunk 4 above)
Comment (rafaelmenta) [reply oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:74]: I think it got back during a rebase. Removed again

File: oauth2_http/javatests/com/google/auth/oauth2/JwtClaimsTest.java (hunk 6 above)
Comment (rafaelmenta) [reply oauth2_http/javatests/com/google/auth/oauth2/JwtClaimsTest.java:37]: fixed

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java (hunk 5 above)
Comment (rafaelmenta) [reply oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:150]: I asked a fellow programmer friends who are more experience than me on Java before coming up with this and they didn't know of any.

I ultimately got inspired by the logic used on [java-jwt](https://github.com/auth0/java-jwt/commit/343c3a2454f889c399ceed755bb5f6d33f2e2496?diff=unified) though I wanted to provide a more robust support than that implementation.

Open to any suggestions you may have on how to improve this code!

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java [hunk 7]
```diff
@@ -111,14 +116,123 @@ public boolean isComplete() {
...
+      return true;
+    }
+
+    /**
+     * Validates if the given object is an instance of a valid JSON basic type.
+     *
+     * @param value object to be evaluated.
+     */
+    private static final boolean isSupportedValue(Object value) {
+      Class c = value.getClass();
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:234]: c --> clazz (google style is no abbreviations)

File: oauth2_http/javatests/com/google/auth/oauth2/JwtClaimsTest.java [hunk 8]
```diff
@@ -33,9 +33,15 @@
 
 import static org.junit.Assert.*;
```
Comment (elharo) [inline oauth2_http/javatests/com/google/auth/oauth2/JwtClaimsTest.java:34]: no wild card imports, per google style

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java (hunk 7 above)
Comment (rafaelmenta) [reply oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:234]: Fixed, thanks

File: oauth2_http/javatests/com/google/auth/oauth2/JwtClaimsTest.java (hunk 8 above)
Comment (rafaelmenta) [reply oauth2_http/javatests/com/google/auth/oauth2/JwtClaimsTest.java]: Fixed, thanks

File: oauth2_http/java/com/google/auth/oauth2/JwtClaims.java [hunk 9]
```diff
@@ -68,10 +73,10 @@
    *
    * @return additional claims
    */
-  abstract Map<String, String> getAdditionalClaims();
+  abstract Map<String, ?> getAdditionalClaims();
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/JwtClaims.java:76]: this is a breaking change. It should be possible to keep the old one marked as Deprecated and calling this new signature internally

**Review by TimurSadykov — CHANGES_REQUESTED** [review]

**Comment by TimurSadykov** [comment]
Closing as outdated, feel free to reopen if actual again
